Bug 1204432 - LazyLoader for SyncEngine, r=ferjm

michielbdejong opened the pull request:
I could do `SyncEngineMock.calledWith` and `SyncEngineMock.shouldFail` in a nice Sinon way, but as it's working this way and it's simple enough, I was hoping maybe we can make that a follow-up bug.

I added a comment (line 33 of apps/sync/js/launcher.js) to reference https://bugzilla.mozilla.org/show_bug.cgi?id=1206615 about adding Kinto.js, but wanted to land this first, since it's nicer to wait for Kinto.js 1.0 next week before we add that.

Commit:
Bug 1204432 - LazyLoader for SyncEngine, r=ferjm

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=ff64f6ad310ef7efec815483e8341c02141303a7

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=202029abd14484cd6173a12e3bb80933e68ad231

ferjm commented on apps/sync/index.html:
```diff
@@ -11,4 +11,6 @@
      </p>
   </body>
+  <script src="shared/js/lazy_loader.js"></script>
+  <script src="js/launcher.js"></script>
```
Maybe rename this to bootstrap to match other apps naming

ferjm commented on apps/sync/js/launcher.js:
```diff
@@ -0,0 +1,81 @@
+      */
+    handleSyncRequest(request) {
+      if (running) {
+        return Promise.reject(new Error('A sync request is already running.'));
```
Add the errors to [1] and follow the same naming convention, please. (i.e. ERROR_SYNC_APP_SYNC_IN_PROGRESS)

[1] https://mxr.mozilla.org/gaia/source/shared/js/sync/errors.js

ferjm replied:
And BTW, this is not supposed to happen as SyncStateMachine takes care of that.

ferjm commented on apps/sync/js/launcher.js:
```diff
@@ -0,0 +1,81 @@
+      if (running) {
+        return Promise.reject(new Error('A sync request is already running.'));
+      }
+
```
Check that `request` has everything that we need here, please.

ferjm commented on apps/sync/js/launcher.js:
```diff
@@ -0,0 +1,81 @@
+          return loadDataAdapterScript(collectionName);
+        }));
+      }).then(() => {
+        var se = new SyncEngine({
```
s/se/engine/g please :)

ferjm commented on apps/sync/test/unit/TestAdapterMock.js:
```diff
@@ -0,0 +1,15 @@
+*/
+
+const TestAdapterMock = {
+  whatAmI: 'the test adapter',
```
Why do you need this?

michielbdejong replied:
Right, I thought Chai `.deep.equal` would need it, but the functions already make the object unique enough. I'm removing it.

ferjm commented on apps/sync/test/unit/lazy-loader-mock.js:
```diff
@@ -0,0 +1,36 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Can you extend and use https://mxr.mozilla.org/gaia/source/shared/test/unit/mocks/mock_lazy_loader.js instead?

ferjm commented on apps/sync/js/launcher.js:
```diff
@@ -0,0 +1,81 @@
+    return LazyLoader.load(`js/adapters/${collectionName}.js`);
+  };
+
+  return {
```
I don't think we need to return anything here. I am adding the IAC handler here.

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=59085166d5e5821b1fd9a7679f2987b2f541424c

michielbdejong commented:
@ferjm Addressed your comments. Shall I rebase and squash?

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=34d2bd8f5b354c31e302c9842fccd58939d298cd

ferjm commented on apps/sync/index.html:
```diff
@@ -11,4 +11,7 @@
        are still in development.
      </p>
   </body>
+  <script src="shared/js/sync/errors.js"></script>
```
We can probably lazy load this script as well. It's only needed in case of error.

ferjm commented on apps/sync/test/unit/bootstrap_test.js:
```diff
@@ -0,0 +1,182 @@
+  requireApp,
+  suite,
+  SyncEngineMock,
+  test
```
nit: alphabetically ordered, please.

ferjm commented on apps/sync/test/unit/bootstrap_test.js:
```diff
@@ -0,0 +1,182 @@
+requireApp('sync/test/unit/sync-engine-mock.js');
+requireApp('sync/test/unit/test-adapter-mock.js');
+require('/shared/test/unit/mocks/mock_lazy_loader.js');
+requireApp('sync/test/unit/lazy-loader-mock.js');
```
Can't you extend shared/test/unit/mocks/mock_lazy_loader.js and get rid of this mock?

ferjm commented on apps/sync/test/unit/TestAdapterMock.js:
```diff
@@ -0,0 +1,14 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
```
Test files are not licensed with MPL2. Use Public Domain instead, please.

https://www.mozilla.org/en-US/MPL/headers/

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=5310c05746e7dc5f31de0486b81ce312ab09ecf0

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=d00507c470e798d11531b343bbf1099169adf976

Commit:
Bug 1209980 - Correct license headers in unit tests, r=ferjm

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f723757fc4bff9078ee4cbb5a7c6222cfeec6a88

michielbdejong commented:
@ferjm Done, squashed and rebased.

I had to fix MockLazyLoader, and found a bug about it; referenced that in https://github.com/michielbdejong/gaia/blob/1204432-sync-app-lazyloader/apps/sync/test/unit/improved_mock_lazy_loader.js.

I also fixed the license header in other unit test files of the sync app, put that in a separate commit as it's unrelated to the main commit of this PR (hope that was the correct way to do it).

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=8825fde9cc4069cb6b46ee534226b1e96fe92408

autolander commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=916785a48f8ce645237b631d4ca83d4085799fc6
